README.md
[![Python Package](https://github.com/Suspir0n/jsonmix/actions/workflows/github-actions-demo.yml/badge.svg)](https://github.com/Suspir0n/jsonmix/actions/workflows/github-actions-demo.yml)
[![PyPI version fury.io](https://badge.fury.io/py/jsonmix.svg)](https://pypi.python.org/pypi/jsonmix/)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/jsonmix.svg)](https://pypi.python.org/pypi/jsonmix/)
[![GitHub license](https://img.shields.io/github/license/Suspir0n/jsonmix.svg)](https://github.com/Suspir0n/jsonmix/blob/main/LICENSE)
[![GitHub tag](https://img.shields.io/github/tag/Suspir0n/jsonmix.svg)](https://github.com/Suspir0n/jsonmix/tags)


# Json Mix

A Python library for validate json. In this `0.1.0` version of jsonmix, it allows you to validate the json that you this awaiting.

## Installation

The latest stable version [is available on PyPI](https://pypi.org/project/jsonmix/). Either add `jsonmix` to your `requirements.txt` file or install with pip:

    pip install jsonmix

## Description

This project is to help other types of projects that need a json validation, and with this lib, you can validate fields, validate the types of fields or if it is necessary to validate both together.

## Steps

### Step 0:

In this lib we use the decorator, so for you to use it would be through a function or class, so it would start by importing the lib:

    import jsonmix

To be more specific and save space in memory, it could also import this way:

    from jsonmix.validate import validate_json

examples of how to use:

* operation receiving fields
  
        def hello():
            receive_json = {
                'uid': 1,
                'Name': 'Json Mix',
                'Version': '0.1.0',
                'message': 'A library for validate in python'
            }
            model_json = {
                'uid': int,
                'Name': str,
                'Version': str,
                'message': str
            }
            response = {
                "code": "JMERR-001",
                "data": {
                    "error": "inconsistency on JSON structure",
                    "message": "Missing required JSON field"
                },
                "date": str(datetime.now().timestamp()),
            }
            return validate_json(receive=receive_json, model=model_json, response_field=response, operation='fields')
    
* operation receiving fields types
  
            def hello():
                receive_json = {
                    'uid': 1,
                    'Name': 'Json Mix',
                    'Version': '0.1.0',
                    'message': 'A library for validate in python'
                }
                model_json = {
                    'uid': int,
                    'Name': str,
                    'Version': str,
                    'message': str
                }
                response = {
                    "code": "JMERR-002",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "JSON field type is incorrect"
                    },
                    "date": str(datetime.now().timestamp())
                }
                return validate_json(receive=receive_json, model=model_json, response_field=response, operation='types')
    
* operation receiving decorator
  
            receive_json = {
                    'uid': 1,
                    'Name': 'Json Mix',
                    'Version': '0.1.0',
                    'message': 'A library for validate in python'
                }
            model_json = {
                'uid': int,
                'Name': str,
                'Version': str,
                'message': str
            }
            response_fields = {
                "code": "JMERR-001",
                "data": {
                    "error": "inconsistency on JSON structure",
                    "message": "Missing required JSON field"
                },
                "date": str(datetime.now().timestamp()),
            }
            response_types = {
                "code": "JMERR-002",
                "data": {
                    "error": "inconsistency on JSON structure",
                    "message": "JSON field type is incorrect"
                },
                "date": str(datetime.now().timestamp())
            }

            @validate_json(receive=receive_json, model=model_json, response_field=response_fields, response_field_types=response_types, operation='decorator')
            def hello():
                return 'Hello Word'

  * operation receiving default
    
            def hello():
                receive_json = {
                        'uid': 1,
                        'Name': 'Json Mix',
                        'Version': '0.1.0',
                        'message': 'A library for validate in python'
                    }
                model_json = {
                    'uid': int,
                    'Name': str,
                    'Version': str,
                    'message': str
                }
                response_fields = {
                    "code": "JMERR-001",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "Missing required JSON field"
                    },
                    "date": str(datetime.now().timestamp()),
                }
                response_types = {
                    "code": "JMERR-002",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "JSON field type is incorrect"
                    },
                    "date": str(datetime.now().timestamp())
                }
                return validate_json(receive=receive_json, model=model_json, response_field=response_fields, response_field_types=response_types, operation='decorator')

Notice that I pass 4 parameters that are them:

1. receive_json: The json you receive from an API or elsewhere.
2. model_json: The model json you created to be the basis of your validations
3. response_json_field: Your error response if the error is in the json keys
4. response_json_field_type: Your error response if the error is in the json types

### Step 2:

We have a test example of each of the parameter passages

* receive_json
```
    example_json_receive = {
        'uid': 1,
        'Name': 'Json Mix',
        'Version': '0.1.0',
        'message': 'A library for validate in python'
    }
```   

* model_json
```
    example_json_model = {
        'uid': int,
        'Name': str,
        'Version': str,
        'message': str
    }
```  

* response_json_field
```
  example_json_custom_error_response_field = {
        "code": "JMERR-001",
        "data": {
            "error": "inconsistency on JSON structure",
            "message": "Missing required JSON field"
        },
        "date": str(datetime.now().timestamp())
  }
```  

* response_json_field_type
```
    example_json_custom_error_response_field_type = {
        "code": "JMERR-002",
        "data": {
            "error": "inconsistency on JSON structure",
            "message": "JSON field type is incorrect"
        },
        "date": str(datetime.now().timestamp())
    }
```  

Now that I have shown you the examples of possible json of each parameter that `validate_json` receives and I have shown an example of use.

### Step 3:

Let's suppose that you set up your json model, to be able to validate with the json that is coming, and this was your model json:

```
    example_json_model = {
        'uid': int,
        'Name': str,
        'Version': str,
        'message': str
    }
``` 

With that there are some malicious people who want to invade or damage your website, app or some software of your own in some way and start editing your receiving json like changing the names of the keys, changing the types or even adding a new key to your json , see these examples of possible json that a malicious person could pass on to you.

```
    example_json_model = {
        'uid': int,
        'Name': str,
        'Version': str,
        'message': str
    }
``` 

```
    example_json_model = {
        'uid': int,
        'Name': str,
        'Version': str,
        'message': str
    }
```

```
    example_json_model = {
        'uid': int,
        'Name': str,
        'Version': str,
        'message': str
    }
``` 

Soon you realize that in the first example you receive a json with the wrong type, in this case your key 'uid' you defined as int and is receiving it as a string, in the second example we can see that an extra key was assigned in your json that would be the 'add_field' that is not in your json model, in the third example you receive a json with the invalid key, in which case your key for the json model is 'Name' and what you received is 'Na'. So these are the possible invalid json that can be received.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- None 

## [0.1.0] - 05/06/2021
### Added
- implemented automated tests
- implemented validation operation types
- implemented tests
- implemented docstring
- implemented the file change log

### Changed
- Some README information.

### Removed
- None

### Fixed
- None

tests/test_validate_json_valid.py
from jsonmix.validate import validate_json
from examples_jsons.valid_json import EJMV_001, EJRV_001, ESJMV_001, ESJRV_001, EJCERF_001, EJCERT_001

# EJRV = example json receive valid
# EJMV = example json model valid
# ESJRV = example strong json receive valid
# ESJMV = example strong json model valid
# EJCERT = example json custom error response types
# EJCERF = example json custom error response fields


def test_validate_json_valid():
    sort = dict
    got = [
        type(validate_json(receive=EJRV_001, model=EJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001)),
        type(validate_json(receive=ESJRV_001, model=ESJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001))
    ]
    for item in got:
        assert verification_condition((item != sort)) is True

    @validate_json(receive=EJRV_001, model=EJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001, operation='decorator')
    def hello():
        _message = 'hello word'
        return _message
    assert verification_condition(type(hello()) != sort) is True


def verification_condition(condition):
    if condition is True:
        return True
    else:
        return False


tests/test_validate_fields_valid.py
from jsonmix.validate import validate_json
from examples_jsons.valid_json import EJMV_001, EJRV_001, ESJMV_001, ESJRV_001, EJCERF_001

# EJRV = example json receive valid
# EJMV = example json model valid
# ESJRV = example strong json receive valid
# ESJMV = example strong json model valid
# EJCERT = example json custom error response types
# EJCERF = example json custom error response fields


def test_validate_fields_valid():
    sort = dict
    got = [
        type(validate_json(receive=EJRV_001, model=EJMV_001, response_field=EJCERF_001, operation='fields')),
        type(validate_json(receive=ESJRV_001, model=ESJMV_001, response_field=EJCERF_001, operation='fields'))
    ]
    for item in got:
        assert verification_condition((item != sort)) is True


def verification_condition(condition):
    if condition is True:
        return True
    else:
        return False


tests/test_validate_json_invalid.py
from jsonmix.validate import validate_json
from examples_jsons.invalid_json import EJRI_001, EJRI_002, EJRI_003, EJRI_004, EJRI_005, ESJRI_003, EJMV_001, ESJMV_001, EJCERF_001, EJCERT_001

# EJRI = example json receive invalid
# EJMV = example json model valid
# ESJRI = example strong json receive invalid
# ESJMV = example strong json model valid
# EJCERT = example json custom error response types
# EJCERF = example json custom error response fields


def test_validate_json_invalid():
    sort = dict
    got = [
        type(validate_json(receive=EJRI_001, model=EJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001)),
        type(validate_json(receive=EJRI_002, model=EJMV_001, response_field=EJCERF_001, operation='fields')),
        type(validate_json(receive=EJRI_003, model=EJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001)),
        type(validate_json(receive=EJRI_004, model=EJMV_001, response_field_type=EJCERT_001, operation='types')),
        type(validate_json(receive=EJRI_005, model=EJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001)),
        type(validate_json(receive=ESJRI_003, model=ESJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001))
    ]
    for item in got:
        assert verification_condition((item != sort)) is False

    @validate_json(receive=ESJRI_003, model=ESJMV_001, response_field=EJCERF_001, response_field_type=EJCERT_001, operation='decorator')
    def hello():
        _message = 'hello word'
        return _message
    assert verification_condition(type(hello()) != sort) is False


def verification_condition(condition):
    if condition is True:
        return True
    else:
        return False


jsonmix/validate.py
from functools import wraps


def _validate_fields(receive:dict, model:dict, response:dict) -> dict:
    """
    This method valid the fields of json receive through of model json passed,
    beyond of pass the receive json and model json, it receives a custom json response.
    :param receive: get the json that comes from outside
    :param model: get the model json
    :param response: get the response customized
    :return: the response customized
    """
    missing = list()
    for key in receive.keys():
        if isinstance(receive[key], dict):
            for i in receive[key].keys():
                if i not in model[key]:
                    missing.append(i)
        if key not in model:
            missing.append(key)
    if missing:
        response.update({'aux': missing})
        return response
    else:
        return None


def _validate_fields_types(receive:dict, model:dict, response:dict) -> dict:
    """
    This method valid the fields types of json receive through of model json passed,
    beyond of pass the receive json and model json, it receives a custom json response.
    :param receive: get the json that comes from outside
    :param model: get the model json
    :param response: get the response customized
    :return: the response customized
    """
    wrong_types = list()
    for type_field in receive.keys():
        if isinstance(receive[type_field], dict):
            for i in receive[type_field].keys():
                if type(receive[type_field][i]) != model[type_field][i]:
                    wrong_types.append(i)
        elif type(receive[type_field]) != model[type_field]:
            wrong_types.append(type_field)
    if wrong_types:
        response.update({'aux': {k: str(v) for k, v in model.items()}})
        return response
    else:
        return None


def validate_json(receive:dict, model:dict, response_field=dict, response_field_type=dict, operation='default') -> dict:
    """
    This method validates json completely, it can only validate fields,
    fields types or both. it receives a json from outside, a valid json
    model and the answer that it will have to have and the type of
    operation, if nothing is passed in the operation parameter, it will
    understand that it has to validate both fields and types, if passing
    "fields" it validates only the fields, if passing "types" it validates
    only the types of the fields, if you pass 'decorator' as the name says,
    you will be able to validate both the fields and the field types, more
    not as a method but as a decorator.

    examples of how to use:

        * operation receiving fields
            def hello():
                receive_json = {
                    'uid': 1,
                    'Name': 'Json Mix',
                    'Version': '0.1.0',
                    'message': 'A library for validate in python'
                }
                model_json = {
                    'uid': int,
                    'Name': str,
                    'Version': str,
                    'message': str
                }
                response = {
                    "code": "JMERR-001",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "Missing required JSON field"
                    },
                    "date": str(datetime.now().timestamp()),
                }
                return validate_json(receive=receive_json, model=model_json, response_field=response, operation='fields')

        * operation receiving fields types
            def hello():
                receive_json = {
                    'uid': 1,
                    'Name': 'Json Mix',
                    'Version': '0.1.0',
                    'message': 'A library for validate in python'
                }
                model_json = {
                    'uid': int,
                    'Name': str,
                    'Version': str,
                    'message': str
                }
                response = {
                    "code": "JMERR-002",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "JSON field type is incorrect"
                    },
                    "date": str(datetime.now().timestamp())
                }
                return validate_json(receive=receive_json, model=model_json, response_field=response, operation='types')

        * operation receiving decorator
            receive_json = {
                    'uid': 1,
                    'Name': 'Json Mix',
                    'Version': '0.1.0',
                    'message': 'A library for validate in python'
                }
            model_json = {
                'uid': int,
                'Name': str,
                'Version': str,
                'message': str
            }
            response_fields = {
                "code": "JMERR-001",
                "data": {
                    "error": "inconsistency on JSON structure",
                    "message": "Missing required JSON field"
                },
                "date": str(datetime.now().timestamp()),
            }
            response_types = {
                "code": "JMERR-002",
                "data": {
                    "error": "inconsistency on JSON structure",
                    "message": "JSON field type is incorrect"
                },
                "date": str(datetime.now().timestamp())
            }

            @validate_json(receive=receive_json, model=model_json, response_field=response_fields, response_field_types=response_types, operation='decorator')
            def hello():
                return 'Hello Word'

        * operation receiving default
            def hello():
                receive_json = {
                        'uid': 1,
                        'Name': 'Json Mix',
                        'Version': '0.1.0',
                        'message': 'A library for validate in python'
                    }
                model_json = {
                    'uid': int,
                    'Name': str,
                    'Version': str,
                    'message': str
                }
                response_fields = {
                    "code": "JMERR-001",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "Missing required JSON field"
                    },
                    "date": str(datetime.now().timestamp()),
                }
                response_types = {
                    "code": "JMERR-002",
                    "data": {
                        "error": "inconsistency on JSON structure",
                        "message": "JSON field type is incorrect"
                    },
                    "date": str(datetime.now().timestamp())
                }
                return validate_json(receive=receive_json, model=model_json, response_field=response_fields, response_field_types=response_types, operation='decorator')


    :param receive: get the json that comes from outside
    :param model: get the model json
    :param response_field: get the response customized for errors in fields
    :param response_field_type: get the response customized for errors in fields types
    :param operation: initialized as 'default', more can get 'fields', 'types' or 'decorator'
    :return: the response customized
    """
    try:
        if operation == 'fields':
            if response_field != '':
                return _validate_fields(receive, model, response_field)
        elif operation == 'types':
            if response_field_type != '':
                return _validate_fields_types(receive, model, response_field_type)
        elif operation == 'decorator':
            def decorator(fn):
                @wraps(fn)
                def wrapper(*args, **kwargs):
                    if response_field != '':
                        missing = _validate_fields(receive, model, response_field)
                        if missing != None:
                            return missing
                    if response_field_type != '':
                        wrong_types = _validate_fields_types(receive, model, response_field_type)
                        if wrong_types != None:
                            return wrong_types
                    return fn(*args, **kwargs)
                return wrapper
            return decorator
        elif operation == 'default':
            if response_field != '':
                missing = _validate_fields(receive, model, response_field)
                if missing != None:
                    return missing
            if response_field_type != '':
                wrong_types = _validate_fields_types(receive, model, response_field_type)
                if wrong_types != None:
                    return wrong_types
    except Exception as error:
        raise Exception(f'The key you passed in the operation parameter is invalid, it only receives "fields", "types" or "decorator". error {error}')




examples_jsons/valid_json.py
from datetime import datetime

# EJRV = example json receive valid
# EJMV = example json model valid
# ESJRV = example strong json receive valid
# ESJMV = example strong json model valid
# EJCERT = example json custom error response types
# EJCERF = example json custom error response fields

EJRV_001 = {
    'uid': 1,
    'Name': 'Json Mix',
    'Version': '0.1.0',
    'message': 'A library for validate in python'
}
EJMV_001 = {
    'uid': int,
    'Name': str,
    'Version': str,
    'message': str
}
ESJRV_001 = {
    'uid': 1,
    'data': {
        'Name': 'Json Mix',
        'Version': '0.1.0',
        'message': 'A library for validate in python'
    },
    'date': str(datetime.now().timestamp())
}
ESJMV_001 = {
    'uid': int,
    'data': {
        'Name': str,
        'Version': str,
        'message': str
    },
    'date': str
}
EJCERT_001 = {
    "code": "JMERR-002",
    "data": {
        "error": "inconsistency on JSON structure",
        "message": "JSON field type is incorrect"
    },
    "date": str(datetime.now().timestamp())
}
EJCERF_001 = {
    "code": "JMERR-001",
    "data": {
        "error": "inconsistency on JSON structure",
        "message": "Missing required JSON field"
    },
    "date": str(datetime.now().timestamp()),
}

examples_jsons/invalid_json.py
from datetime import datetime

# EJRI = example json receive invalid
# EJMV = example json model valid
# ESJRI = example strong json receive invalid
# ESJMV = example strong json model valid
# EJCERT = example json custom error response types
# EJCERF = example json custom error response fields

EJRI_001 = {
    'uid': '1',
    'Na': 'Json Mix',
    'Vesion': '0.1.0',
    'message': 'A library for validate in python'
}
EJRI_002 = {
    'uid': '1',
    'Name': 'Json Mix',
    'Version': '0.1.0',
    'message': 'A library for validate in python',
    'add_field': 'test'
}
EJRI_003 = {
    'uid': '1',
    'Name': 'Json Mix',
    'add_field': 'test',
    'Version': '0.1.0',
    'message': 'A library for validate in python'
}
EJRI_004 = {
    'uid': '1',
    'Name': 'Json Mix',
    'Version': '0.1.0',
    'message': 'A library for validate in python'
}
EJRI_005 = {
    'uid': '1',
    'Name': 'Json Mix',
    'Version': 0.1,
    'message': 'A library for validate in python'
}
ESJRI_001 = {
    'uid': '1',
    'data': {
        'Na': 'Json Mix',
        'Version': '0.1.0',
        'message': 'A library for validate in python'
    },
    'date': str(datetime.now().timestamp())
}
ESJRI_002 = {
    'uid': '1',
    'data': {
        'Name': 'Json Mix',
        'Version': '0.1.0',
        'message': 'A library for validate in python'
    },
    'date': datetime.now().timestamp()
}
ESJRI_003 = {
    'uid': '1',
    'data': {
        'Nam': 'Json Mix',
        'Version': '0.1.0',
        'message': 'A library for validate in python'
    },
    'date': datetime.now().timestamp()
}
ESJMV_001 = {
    'uid': int,
    'data': {
        'Name': str,
        'Version': str,
        'message': str
    },
    'date': str
}
EJMV_001 = {
    'uid': int,
    'Name': str,
    'Version': str,
    'message': str
}
EJCERF_001 = {
    "code": "JMERR-001",
    "data": {
        "error": "inconsistency on JSON structure",
        "message": "Missing required JSON field",
    },
    "date": str(datetime.now().timestamp()),
}
EJCERT_001 = {
    "code": "JMERR-002",
    "data": {
        "error": "inconsistency on JSON structure",
        "message": "JSON field type is incorrect"
    },
    "date": str(datetime.now().timestamp())
}
